wpimath/src/main/native/include/frc/kinematics/Odometry3d.h

```diff
@@ -0,0 +1,237 @@
+   *
+   * @param gyroAngle The angle reported by the gyroscope.
+   * @param wheelPositions The current distances measured by each wheel.
+   * @param pose The position on the field that your robot is at.
+   */
+  void ResetPosition(const Rotation2d& gyroAngle,
+                     const WheelPositions& wheelPositions, const Pose2d& pose) {
+    ResetPosition(Rotation3d{0_rad, 0_rad, gyroAngle.Radians()}, wheelPositions,
+                  Pose3d{pose});
+  }
+
+  /**
+   * Resets the robot's position on the field.
+   *
+   * The gyroscope angle does not need to be reset here on the user's robot
+   * code. The library automatically takes care of offsetting the gyro angle.
+   *
+   * @param gyroAngle The angle reported by the gyroscope.
+   * @param wheelPositions The current distances measured by each wheel.
+   * @param pose The position on the field that your robot is at.
+   */
+  void ResetPosition(const Rotation3d& gyroAngle,
+                     const WheelPositions& wheelPositions, const Pose3d& pose) {
+    m_pose = pose;
+    m_pose2d = m_pose.ToPose2d();
+    m_previousAngle = pose.Rotation();
+    m_gyroOffset = m_pose.Rotation() - gyroAngle;
+    m_previousWheelPositions = wheelPositions;
+  }
+
+  /**
+   * Resets the pose.
+   *
+   * @param pose The pose to reset to.
+   */
+  void ResetPose(const Pose2d& pose) { ResetPose(Pose3d{pose}); }
+
+  /**
+   * Resets the pose.
+   *
+   * @param pose The pose to reset to.
+   */
+  void ResetPose(const Pose3d& pose) {
+    m_gyroOffset = m_gyroOffset + (pose.Rotation() - m_pose.Rotation());
+    m_pose = pose;
+    m_pose2d = m_pose.ToPose2d();
+    m_previousAngle = pose.Rotation();
+  }
+
+  /**
+   * Resets the translation of the pose.
+   *
+   * @param translation The translation to reset to.
+   */
+  void ResetTranslation(const Translation2d& translation) {
+    ResetTranslation(Translation3d{translation.X(), translation.Y(), 0_m});
+  }
+
+  /**
+   * Resets the translation of the pose.
+   *
+   * @param translation The translation to reset to.
+   */
+  void ResetTranslation(const Translation3d& translation) {
+    m_pose = Pose3d{translation, m_pose.Rotation()};
+    m_pose2d = m_pose.ToPose2d();
+  }
+
+  /**
+   * Resets the rotation of the pose.
+   *
+   * @param rotation The rotation to reset to.
+   */
+  void ResetRotation(const Rotation2d& rotation) {
+    ResetRotation(Rotation3d{0_rad, 0_rad, rotation.Radians()});
+  }
+
+  /**
+   * Resets the rotation of the pose.
+   *
+   * @param rotation The rotation to reset to.
+   */
+  void ResetRotation(const Rotation3d& rotation) {
+    m_gyroOffset = m_gyroOffset + (rotation - m_pose.Rotation());
+    m_pose = Pose3d{m_pose.Translation(), rotation};
+    m_pose2d = m_pose.ToPose2d();
+    m_previousAngle = rotation;
+  }
+
+  /**
+   * Returns the position of the robot on the field.
+   * @return The pose of the robot.
+   */
+  const Pose2d& GetPose() const { return m_pose2d; }
```

Comment:
Should Java also cache the 2D pose?

Reply:
Probably. The main point of the cache is really so that the reference is valid, but it wouldn't hurt to make the implementations more consistent and save on allocations. There also shouldn't be any wasted work in updating the cache, since it's likely that every call to update will also want the 2d pose.
By the way, I'm at school, so sometimes I'll be unresponsive for an hour or two.